packages/lodestar-beacon-state-transition/src/phase0/fast/util/cachedBeaconState.ts

```diff
@@ -0,0 +1,105 @@
+import {ContainerType, ITreeBacked, List, readOnlyMap, TreeBacked} from "@chainsafe/ssz";
+import {allForks} from "@chainsafe/lodestar-types";
+import {IBeaconConfig} from "@chainsafe/lodestar-config";
+import {Tree} from "@chainsafe/persistent-merkle-tree";
+import {MutableVector} from "@chainsafe/persistent-ts";
+import {createFlat} from "./flat";
+import {createEpochContext, EpochContext} from "./epochContext";
+import {CachedValidatorList, CachedValidatorListProxyHandler} from "./cachedValidatorList";
+
+/**
+ * `BeaconState` with various caches
+ *
+ * Currently contains the following:
+ * - The full list of network params, ssz types, and fork schedule
+ * - The ssz type for the state
+ * - The full merkle tree representation of the state
+ * - A cache of shufflings, committees, proposers, expanded pubkeys
+ * - A flat copy of validators (for fast access/iteration)
+ */
+export type CachedBeaconState<T extends allForks.BeaconState> =
+  // Wrapper object ({validators, clone()})
+  BeaconStateContext<T> &
+    // Epoch Cache
+    EpochContext &
+    // SSZ ops
+    ITreeBacked<T> &
+    // Beacon State interface
+    T;
+
+export function createCachedBeaconState<T extends allForks.BeaconState>(
+  config: IBeaconConfig,
+  state: TreeBacked<T>
+): CachedBeaconState<T> {
+  const cachedValidators = MutableVector.from(readOnlyMap(state.validators, (v) => createFlat(v)));
+  const epochCtx = createEpochContext(config, state, cachedValidators);
+  return new Proxy(
+    new BeaconStateContext(state, cachedValidators, epochCtx),
+    (CachedBeaconStateProxyHandler as unknown) as ProxyHandler<BeaconStateContext<T>>
+  ) as CachedBeaconState<T>;
+}
+
+export class BeaconStateContext<T extends allForks.BeaconState> {
+  config: IBeaconConfig;
+  // epoch cache
+  epochCtx: EpochContext;
+  // the BeaconState ssz type
+  type: ContainerType<T>;
+  // the original BeaconState as a Tree
+  tree: Tree;
+  // immutable and shared across BeaconStates for most of the validators
+  protected _validatorCache: MutableVector<T["validators"][number]>;
+
+  constructor(state: TreeBacked<T>, validatorCache: MutableVector<T["validators"][number]>, epochCtx: EpochContext) {
+    this.config = epochCtx.config;
+    this.type = state.type() as ContainerType<T>;
+    this.tree = state.tree();
+    this.epochCtx = epochCtx;
+    this._validatorCache = validatorCache;
+  }
+
+  get validators(): CachedValidatorList<T["validators"][number]> & T["validators"] {
+    return (new Proxy(
+      new CachedValidatorList(
+        (this.type.tree.getProperty(this.tree, "validators") as unknown) as TreeBacked<List<T["validators"][number]>>,
```

Comment:
not for this Pr but we could maybe add optional generic to getProperty to make it easier to set expected type:
```
this.type.tree.getProperty<TreeBacked<List<T["validators"][number]>>>(this.tree, "validators")
```